src/Raven.Server/Documents/AbstractBackgroundWorkStorage.cs

```diff
@@ -183,25 +189,36 @@ public static bool HasPassed(BlittableJsonReaderObject metadata, DateTime curren
     
     protected abstract void ProcessDocument(DocumentsOperationContext context, Slice lowerId, string id, DateTime currentTime);
 
-    public int ProcessDocuments(DocumentsOperationContext context, Dictionary<Slice, List<(Slice LowerId, string Id)>> docsToProcess, DateTime currentTime)
+    public int ProcessDocuments(DocumentsOperationContext context, Queue<DocumentExpirationInfo> expired, DateTime currentTime)
     {
         var processedCount = 0;
-        var docsTree = context.Transaction.InnerTransaction.ReadTree(_treeName);
+        var count = 0;
 
-        foreach (var pair in docsToProcess)
+        var docsTree = context.Transaction.InnerTransaction.ReadTree(_treeName);
+        foreach (var info in expired)
         {
-            foreach (var ids in pair.Value)
+            if (info.Id != null)
             {
-                if (ids.Id != null)
-                {
-                    ProcessDocument(context, ids.LowerId, ids.Id, currentTime);
-                    processedCount++;
-                }
-
-                docsTree.MultiDelete(pair.Key, ids.LowerId);
+                ProcessDocument(context, info.LowerId, info.Id, currentTime);
+                processedCount++;
             }
+
+            count++;
+            docsTree.MultiDelete(info.Ticks, info.LowerId);
         }
 
+        var tx = context.Transaction.InnerTransaction.LowLevelTransaction;
+        tx.OnDispose += _ =>
+        {
+            if (tx.Committed == false)
+                return;
+
+            for (int i = 0; i < count; i++)
```

Comment:
hmm... I think we may iterate while dequeuing
it depends what's more readable for you
```
while (queue.Count > 0)
{
    int element = queue.Dequeue();
```

Reply:
I think that it's more readable to dequeue the `dequeueCount.